### ISSUE TITLE
Calling Polly's ExecuteAndCapture<TResult> after migrating to .NetCore 2.1

### ISSUE THREAD

**Issue by AJPoulter** [App-vNext/Polly#515]
**Summary: What are you wanting to achieve?**

After migrating my .NetCore 2.0 Polly implementation to .NetCore 2.1 making use of the new HttpClientFactory and Polly extensions, I want to be able to call the ExecuteAndCapture method in my Data Access Layer classes on the policy that was created and configured in StartUp.cs during service configuration.

**What code or approach do you have so far?**  

In the 2.0 implementation I created a PolicyWrap as a private member (_responseWrapper) in an abstract base class that is made available to derived classes of the DAL in the following method:

protected Task<PolicyResult<HttpResponseMessage>> GetResponsePolicyResultAsync(Func<Task<HttpResponseMessage>> action) => _responseWrapper.ExecuteAndCaptureAsync(() => action());

The aim, of course, is to allow all derived classes to capture and evaluate the result of invoking the HttpClient via the policy(wrap) for the purposes of logging and exception handling.

n the 2.1 scenario with the new HttpClientFactory and the fluent configuration of HttpClient I have been attracted by the out-of-the-box policy overload AddTransientHttpErrorPolicy(builder => builder.WaitAndRetryAsync….

But this is all set up in StartUp.cs so I do not see any way to access this policy in my DAL classes. 

Does this convenient fluent configuration come at the cost that one can no longer call the

 .ExecuteAndCapture<TResult>(Func<TResult>)method of the default policy or policy wrap used to enclose all HttpClient outgoing calls?

**Comment by reisenberger** [comment]
@AJPoulter Yes, the [use of a Polly policy in `DelegatingHandler` middleware of `HttpClient`](https://www.stevejgordon.co.uk/httpclientfactory-aspnetcore-outgoing-request-middleware-pipeline-delegatinghandlers) - which is what the .NET Core 2.1 `HttpClientFactory` extensions configure -  is fundamentally incompatible with `.ExecuteAndCaptureAsync()`. And this is not something the Polly team can change, unfortunately.  

The reason is that extending `DelegatingHandler` requires overriding a [method of signature](https://github.com/aspnet/HttpClientFactory/blob/4dfcf7485e62819d0d73c00d7e7eb53d75abdbd3/src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs#L108) 

    async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) 

If the [call through the Polly policy](https://github.com/aspnet/HttpClientFactory/blob/4dfcf7485e62819d0d73c00d7e7eb53d75abdbd3/src/Microsoft.Extensions.Http.Polly/PolicyHttpMessageHandler.cs#L130) was changed to `.ExecuteAndCaptureAsync()`, that would change the return type to `Task<PolicyResult<HttpResponseMessage>>` - breaking the signature of `SendAsync(...)` and thus not correctly overriding `DelegatingHandler`.

(It would only be possible if Microsoft chose to implement another outgoing codepath through `HttpClient` based on Polly's `PolicyResult<T>`, ie tightly coupled to Polly at the `HttpClient` level, which I don't expect to happen.)

---

What options does that leave for your case?  If you already have a good solution with `PolicyWrap` and `.ExecuteAndCaptureAsync(...)`, it may be as well to stick with that.  If you want to take advantage of the out-of-box definition of `.AddTransientHttpErrorPolicy(...)` (ie the definition of which faults that handles), then we provide that separately from HttpClientFactory, [here](https://www.nuget.org/packages/Polly.Extensions.Http/) as we document [here](https://github.com/App-vNext/Polly/wiki/Polly-and-HttpClientFactory#extending-the-convenience-addtransienthttperrorpolicy-definition).  The caveat with using Polly policies on calls through `HttpClient` _without_ taking the `DelegatingHandler` approach is that retries can fall foul of the exception that an `HttpRequestMessage` cannot be re-used once sent.

If you want to adapt your existing DAL code to a .NET Core DI -style approach, configuring policies in `StartUp.cs`, then that is fairly trivial to do.  Configure a [`PolicyRegistry`](https://github.com/App-vNext/Polly/wiki/PolicyRegistry) and add policies to it during `StartUp.cs`, as [exemplified here](https://github.com/App-vNext/Polly/wiki/Polly-and-HttpClientFactory#selecting-policies-from-a-policyregistry).  Also register an [`IReadOnlyPolicyRegistry`](https://github.com/App-vNext/Polly/blob/7eaf7414256ce6e7a86dc60742e3b0e36d893722/src/Polly.Shared/Registry/IReadOnlyPolicyRegistry.cs) or `IReadOnlyPolicyRegistry<T>` during startup, have your DAL layer receive that `IReadOnlyPolicyRegistry<T>` by DI, and the DAL layer then extracts the policy it needs from the `IReadOnlyPolicyRegistry<T>`.

Hope that helps.

**Comment by AJPoulter** [comment]
Dear Dylan,
Thanks for your prompt and very detailed reply. I understand the issue now and I will adopt your proposal of injecting the PolicyRegistry with the configured Policy into the DAL base class, That covers all the bases for me.
THANKS!
Kind regards

Andrew Poulter
ajpoulter1961@gmail.com
+43 (0)660 457 8017
